File: day3_strategy/guru/py/main.py
from context import *
from strategy import *


# The client code picks a concrete strategy and passes it to the context.
# The client should be aware of the differences between strategies in order
# to make the right choice.
def client_code() -> None:
    context = Context(ConcreteStrategyA())
    print("Client: Strategy is set to normal sorting.")
    context.do_some_business_logic()
    print()

    context.strategy = ConcreteStrategyB()
    print("Client: Strategy is set to reverse sorting.")
    context.do_some_business_logic()


if __name__ == "__main__":
    client_code()
    # Client: Strategy is set to normal sorting.
    # Context: Sorting data using the strategy... (not sure how it'll do it)
    # Result: a,b,c,d,e
    #
    # Client: Strategy is set to reverse sorting.
    # Context: Sorting data using the strategy... (not sure how it'll do it)
    # Result: e,d,c,b,a

File: day3_strategy/guru/main.cpp
#include <iostream>

#include "context.hpp"
#include "strategy.hpp"

// The client code picks a concrete strategy and passes it to the context. The
// client should be aware of the differences between strategies in order to make
// the right choice.
void ClientCode() {
  Context context(new ConcreteStrategyA());
  std::cout << "Client: Strategy is set to normal sorting." << '\n';
  context.DoSomeBusinessLogic();
  std::cout << '\n';

  context.SetStrategy(new ConcreteStrategyB());
  std::cout << "Client: Strategy is set to reverse sorting." << '\n';
  context.DoSomeBusinessLogic();
}


int main(int argc, char const *argv[]) {
  ClientCode();
  // Client: Strategy is set to normal sorting.
  // Context: Sorting data using the strategy... (not sure how it'll do it)
  // Result: abcde
  //
  // Client: Strategy is set to reverse sorting.
  // Context: Sorting data using the strategy... (not sure how it'll do it)
  // Result: edcba
  
  return 0;
}
